### PR TITLE
All DataAdapters can now create a native iterator for each backend.

### PR DESCRIPTION
**Pull request by hertschuh**
- Added `get_jax_iterator` and `get_torch_dataloader` to all `DataAdapter`s.
- `GeneratorDataAdapter` and `PyDatasetAdapter` can now consume tensors from any backend (added support for JAX and Torch). As a result, any combination of input format is supported by these two `DataAdapter`s.
- Made `DataAdapter`s unit tests similar.
- Fixed gap where `shuffle="batch"` was not implemented (used a global shuffle) for the numpy iterator.
- `GeneratorDataAdapter` no longer peeks twice at the first element.
- Removed the concept of `return_type` in `EpochIterator` since it is now always "auto".
- Each backend has a subclass of `EpochIterator` (this is not new), which is now in charge of retrieving the correct iterator for the backend.
- This prevents the double conversion that was happening in some cases (e.g. `TFDatasetAdapter` from `tf.Tensor` to numpy to JAX or Torch). Note that `ArrayDataAdapter` still needs some work to not have double conversions.
- The optimization of using `np.asarray` instead of `np.array` was moved from the `DataAdapter`s to the backend's `convert_to_numpy` since that is what is now used by the `DataAdapter`s.
- Also fixes https://github.com/keras-team/keras/issues/19038 .

### CODE REVIEW

File: keras/backend/torch/trainer.py [hunk 1]
```diff
@@ -496,40 +493,5 @@ def predict_on_batch(self, x):
 
 
 class TorchEpochIterator(EpochIterator):
-    def _get_iterator(self, return_type="auto"):
-        if return_type == "auto" and isinstance(
-            self.data_adapter, data_adapters.TorchDataLoaderAdapter
-        ):
-            return self.data_adapter.get_torch_dataloader()
```
Comment (hertschuh) [inline keras/backend/torch/trainer.py:503]: @james77777778 

The iterator is now always a Torch `DataLoader`. So the prefetching code is ineffective as-is. However, I wonder if I should just move this code.

Do `DataLoader`s do the prefetching for you? There is a `prefetch_factor` argument, which defaults to 2, but it is off if multiprocessing is not on.

File: keras/trainers/data_adapters/array_data_adapter_test.py [hunk 2]
```diff
@@ -1,36 +1,39 @@
...
         elif array_type == "pandas":
-            return pandas.DataFrame(np.ones(shape, dtype=dtype))
+            return pandas.DataFrame(x)
 
-    @parameterized.parameters([("np",), ("tf",), ("backend",), ("pandas",)])
-    def test_basic_flow(self, array_type):
+    @parameterized.named_parameters(
+        named_product(
+            array_type=["np", "tf", "jax", "torch", "pandas"],
+            iterator_type=["np", "tf", "jax", "torch"],
```
Comment (hertschuh) [inline keras/trainers/data_adapters/array_data_adapter_test.py:32]: With this PR, all iterator types are tested with all backends (to continue testing the way it was tested before).

To speed this up a little bit, we could only test the iterator for the current backend. The CI will run the test with all backends anyway. Additionally, the only use case is getting the iterator for the current backend.

File: keras/backend/torch/trainer.py (hunk 1 above)
Comment (fchollet) [reply keras/backend/torch/trainer.py:503]: 1. What is the performance impact of the change? 
2. Can we turn on multiprocessing by default?

Comment (james77777778) [reply keras/backend/torch/trainer.py:503]: Hi @hertschuh 
I believe torch's `DataLoader` will enable prefetching if `num_workers>0`.
Ref: https://pytorch.org/docs/stable/data.html#torch.utils.data.DataLoader

However, it is set to `0` in `get_torch_dataloader` in `data_adapter_utils` which might disable prefetching.

Comment (hertschuh) [reply keras/backend/torch/trainer.py:503]: 1. I tested as-is using the same script as https://github.com/keras-team/keras/pull/18991 (thanks @james77777778  !) on a Tesla V100:

| Code tested | Training speed |
|-------------|----------------|
| **master** (prefetching) | 230-238ms/step |
| **this code** (DataLoader) | 230-238ms/step | 

So there is no change in performance. Additionally, I tried re-adding the `_prefetch_numpy_iterator` logic before feeding it to the `DataLoader`, and that makes it a lot slower, although I don't understand why.

2. I tried that (setting `num_workers=1`) and it fails. Per the error message, in order to do that you have to actually manually spawn new separate processes (even for `num_workers=1`), which would add a lot of complexity.

Comment (hertschuh) [reply keras/backend/torch/trainer.py:503]: As it turns out, there were performance regressions for JAX. They are now fixed (and benchmarked).

The reason why there was no regression for Torch even though the prefetching code was removed is that the implementation for the prefetching used a synchronous function to transfer from CPU to GPU. In comparison, the JAX implementation uses an asynchronous function. I'm playing with some asynchronous way of transferring tensors from CPU to GPU for Torch, but it's not quite working right now (it's much slower for some reason), so I'll address that in a separate PR.

Here are the results (very reproducible, numbers are within ± 1ms).

Notes:
- I tweaked the model size so that the preprocessing on CPU for a step takes about the same time as one step of training on GPU to make the lack of prefetching most detectable (see code below).
- Only the TFDataset benchmark uses preprocessing.

### JAX

| DataAdapter type|Arrays|Generator|PyDataset|TFDataset|DataLoader|
|------------------|---------|---------|---------|---------|---------|
| before this PR      | 222 ms | 222 ms | 222 ms | 259 ms | 244 ms |
| after this PR         | 222 ms | 222 ms | 222 ms | 259 ms | 244 ms |

(no changes)

### Torch

| DataAdapter type|Arrays|Generator|PyDataset|TFDataset|DataLoader|
|------------------|---------|---------|---------|---------|---------|
| before this PR      | 366 ms | 366 ms | 366 ms | 384 ms | 359 ms |
| after this PR         | 353 ms | 366 ms | 366 ms | 384 ms | 359 ms |

`ArrayDataAdapter` is bit faster now because it has a native Torch implementation.

### Test code

Based on @james77777778 's test code in https://github.com/keras-team/keras/pull/18991 but I added coverage for all `DataAdapter` types.

```python
import numpy as np

# isort: off
import torch  # noqa:F401 dummy import to prevent segmentation fault
import tensorflow as tf
# isort: on

import keras
from keras import backend
from keras import layers
from keras import losses
from keras import optimizers
from keras.trainers.data_adapters.py_dataset_adapter import PyDataset

class TestModel(keras.Model):
    def __init__(self, num_layers=4, **kwargs):
        super().__init__(**kwargs)
        self.num_layers = num_layers
        self.convs = []
        for _ in range(num_layers):
            self.convs.append(
                layers.Conv2D(32, 3, padding="same", activation="relu")
            )
        self.pool = layers.GlobalAveragePooling2D()
        self.fatten = layers.Flatten()
        self.dense = layers.Dense(1)

    def call(self, x):
        for i in range(self.num_layers):
            x = self.convs[i](x)
        x = self.pool(x)
        x = self.fatten(x)
        x = self.dense(x)
        return x

def preprocess(x, y):
    x = tf.image.random_brightness(x, 0.2)
    x = tf.image.random_contrast(x, 0.2, 0.5)
    x = tf.image.random_saturation(x, 0.2, 0.5)
    x = tf.image.rgb_to_hsv(x)
    x = tf.image.hsv_to_rgb(x)
    return x, y

model = TestModel(num_layers=8)
model.compile(
    optimizer=optimizers.SGD(learning_rate=0.001),
    loss=losses.MeanSquaredError(),
    jit_compile=False if backend.backend() == "torch" else True,
)

x = np.random.random((1024, 416, 416, 3)).astype("float32")
y = np.random.random((1024, 1)).astype("float32")

print("=== fit with Arrays ===")
model.fit(x, y, batch_size=32, epochs=3)

print("=== fit with Generator ===")
def generate_data():
    while True:
        for idx in range(32):
            low = idx * 32
            high = low + 32
            yield x[low:high], y[low:high]

model.fit(generate_data(), steps_per_epoch=32, epochs=3)

print("=== fit with PyDataset ===")
class ExamplePyDataset(PyDataset):
    def __len__(self):
        return 32

    def __getitem__(self, idx):
        low = idx * 32
        high = low + 32
        return x[low:high], y[low:high]

model.fit(ExamplePyDataset(), epochs=3)

print("=== fit with TFDataset ===")
tf_dataset = tf.data.Dataset.from_tensor_slices((x, y))
tf_dataset = tf_dataset.batch(32)
tf_dataset = tf_dataset.map(preprocess, num_parallel_calls=tf.data.AUTOTUNE)
tf_dataset = tf_dataset.prefetch(tf.data.AUTOTUNE)
model.fit(tf_dataset, epochs=3)

print("=== fit with Torch DataLoader ===")
torch_dataset = torch.utils.data.TensorDataset(torch.as_tensor(x), torch.as_tensor(y))
torch_dataloader = torch.utils.data.DataLoader(torch_dataset, batch_size=32)
model.fit(torch_dataloader, epochs=3)
```

**Review by fchollet — APPROVED** [review]
LGTM, thank you!
